### ISSUE TITLE
Use of scala.reflect in statusMapping causes sporadic failures

### ISSUE THREAD

**Issue by jatcwang** [softwaremill/tapir#371]
scala.reflect API isn't thread safe according to https://github.com/scala/bug/issues/10766 and I'm getting sporadic test failures when I use statusMapping:

```
...
.out(oneOf(statusMapping(StatusCode.Created, jsonBody[MyClass])))
...
```

where `MyClass` is just a normal concrete class `case class MyClass(a: Int)`.

**Comment by jatcwang** [comment]
I'm wondering whether we can add macros to generate instances of `MatchType` for a normal class without type parameters like `MyClass` above (so erasure check is done at compile time instead of runtime). The macro will need to handle `AnyVal`s

**Comment by adamw** [comment]
Now that's surprising ... as I understand, the problem is with `=:=` in the `if (!(t =:= t.erasure)) {` check in `statusMapping`?

One way of  making this thread-safe is running this at compile-time: that's single-threaded, and we should be able to do the check already then. That would be even better for error reporting.

The same check is also used in the `MatchType` magnolia implementation, that would need to be fixed as well.

**Comment by adamw** [comment]
Fixed in 0.12.14.
